crates/wdk-build/src/cargo_make.rs

```diff
@@ -0,0 +1,629 @@
+//! cargo commands. It uses a combination of `clap` and `clap_cargo` to provide
+//! a CLI very close to cargo's own, but only exposes the arguments supported by
+//! `rust-driver-makefile.toml`. Help text and other `clap::Arg`
+
+use std::path::{Path, PathBuf};
+
+use clap::{Args, Parser};
+
+use crate::{
+    utils::{detect_wdk_content_root, get_latest_windows_sdk_version, PathExt},
+    CPUArchitecture,
+    ConfigError,
+};
+
+const PATH_ENV_VAR: &str = "Path";
+
+/// The name of the environment variable that cargo-make uses during `cargo
+/// build` and `cargo test` commands
+const CARGO_MAKE_CARGO_BUILD_TEST_FLAGS_ENV_VAR: &str = "CARGO_MAKE_CARGO_BUILD_TEST_FLAGS";
+
+const CARGO_MAKE_PROFILE_ENV_VAR: &str = "CARGO_MAKE_PROFILE";
+const CARGO_MAKE_CARGO_PROFILE_ENV_VAR: &str = "CARGO_MAKE_CARGO_PROFILE";
+const CARGO_MAKE_CRATE_CUSTOM_TRIPLE_TARGET_DIRECTORY_ENV_VAR: &str =
+    "CARGO_MAKE_CRATE_CUSTOM_TRIPLE_TARGET_DIRECTORY";
+const CARGO_MAKE_RUST_DEFAULT_TOOLCHAIN_ENV_VAR: &str = "CARGO_MAKE_RUST_DEFAULT_TOOLCHAIN";
+const CARGO_MAKE_CRATE_FS_NAME_ENV_VAR: &str = "CARGO_MAKE_CRATE_FS_NAME";
+const WDK_BUILD_OUTPUT_DIRECTORY_ENV_VAR: &str = "WDK_BUILD_OUTPUT_DIRECTORY";
+
+/// `clap` uses an exit code of 2 for usage errors: <https://github.com/clap-rs/clap/blob/14fd853fb9c5b94e371170bbd0ca2bf28ef3abff/clap_builder/src/util/mod.rs#L30C18-L30C28>
+const CLAP_USAGE_EXIT_CODE: i32 = 2;
+
+trait ParseCargoArg {
+    fn parse_cargo_arg(&self);
+}
+
+#[derive(Parser, Debug)]
+struct CommandLineInterface {
+    #[command(flatten)]
+    base: BaseOptions,
+
+    #[command(flatten)]
+    #[command(next_help_heading = "Package Selection")]
+    workspace: clap_cargo::Workspace,
+
+    #[command(flatten)]
+    #[command(next_help_heading = "Feature Selection")]
+    features: clap_cargo::Features,
+
+    #[command(flatten)]
+    compilation_options: CompilationOptions,
+
+    #[command(flatten)]
+    manifest_options: ManifestOptions,
+}
+
+#[derive(Args, Debug)]
+struct BaseOptions {
+    #[arg(long, help = "Do not print cargo log messages")]
+    quiet: bool,
+
+    #[arg(short, long, action = clap::ArgAction::Count, help = "Use verbose output (-vv very verbose/build.rs output)")]
+    verbose: u8,
+}
+
+#[derive(Args, Debug)]
+#[command(next_help_heading = "Compilation Options")]
+struct CompilationOptions {
+    #[arg(
+        short,
+        long,
+        help = "Build artifacts in release mode, with optimizations"
+    )]
+    release: bool,
+
+    #[arg(
+        long,
+        value_name = "PROFILE-NAME",
+        help = "Build artifacts with the specified profile"
+    )]
+    profile: Option<String>,
+
+    #[arg(
+        short,
+        long,
+        value_name = "N",
+        allow_negative_numbers = true,
+        help = "Number of parallel jobs, defaults to # of CPUs."
+    )]
+    jobs: Option<String>,
+
+    // TODO: support building multiple targets at once
```

Comment:
Support for building multiple targets in a single cargo make invocation should be added in the future